Apps/FlightSoftware/Watchdog2/src/ground_cmd.c

```diff
@@ -0,0 +1,498 @@
+        /* malformed packet length */
+        reply_ground_cmd(0, GNDRESP_EPACKETLEN);
+        return;
+    }
+
+    /* TODO: check the checksum */
+    (void)(buf[3]);
+
+    /* get the type (should be 000BADA55 for hercules) */
+    if (buf[4] == 0xEE && buf[5] == 0xFF && buf[6] == 0x00 && buf[7] == 0xC0) {
+        /* command */
+        handle_ground_cmd(buf + 8, pp_len - 8);
+    } else {
+        /* forward it on to hercules */
+        // copy the buffer into the hercules buffer
+        memcpy(hercbuf.buf + hercbuf.used, pp->buf, pp->used); // @suppress("Invalid arguments")
+        hercbuf.used += pp->used;
+        if (hercbuf.used > BUFFER_SIZE) {
+            // TODO: better logic here. should overwrite old like a ring buffer, not just
+            // reset from 0
+            hercbuf.used = 0;
+        }
+    }
+
+    /* done parsing */
+    // TODO: multiple packets? handle.
+    pp->used = 0;
+    pp->idx = 0;
+}
+
+/**
+ * Send the earth heartbeat
+ */
+void send_earth_heartbeat() {
+    static uint8_t counter = 0;
+    uint8_t send_buf[32];
+    if (counter % 3 != 2) {
+        // send every 2 seconds
+        counter++;
+        //return;
+    }
+    counter = 0;
+
+    // build the packet
+    send_buf[0] = 0xFF;
+    // TODO: tvac changes
+    // send adc value temperature
+    send_buf[1] = (uint8_t)(adc_values[ADC_TEMP_IDX]);
+    send_buf[2] = (uint8_t)(adc_values[ADC_TEMP_IDX] >> 8);
+
+    // send adc value temperature
+    send_buf[3] = (uint8_t)(raw_battery_charge[0]);
+    send_buf[4] = (uint8_t)(raw_battery_charge[1]);
+
+    // send adc value temperature
+    send_buf[5] = (uint8_t)(raw_battery_voltage[0]);
+    send_buf[6] = (uint8_t)(raw_battery_voltage[1]);
+
+    // send adc value temperature
+    send_buf[7] = (uint8_t)(raw_battery_current[0]);
+    send_buf[8] = (uint8_t)(raw_battery_current[1]);
+
+    // send adc value temperature
+    send_buf[9] = (uint8_t)(raw_fuel_gauge_temp[0]);
+    send_buf[10] = (uint8_t)(raw_fuel_gauge_temp[1]);
+
+    // send ASDF
+    send_buf[11] = (uint8_t)(Kp_heater);
+    send_buf[12] = (uint8_t)(Kp_heater >> 8);
+
+    // send ASDF
+    send_buf[13] = (uint8_t)(heater_setpoint);
+    send_buf[14] = (uint8_t)(heater_setpoint >> 8);
+
+    // send ASDF
+    send_buf[15] = (uint8_t)(heater_window);
+    send_buf[16] = (uint8_t)(heater_window >> 8);
+
+    // send ASDF
+    send_buf[17] = (uint8_t)(PWM_limit);
+    send_buf[18] = (uint8_t)(PWM_limit >> 8);
+
+    // send ASDF
+
+    // send the current heating status
+    send_buf[19] = 0;
+    send_buf[20] = heaterStatus;
+    send_buf[21] = heatingControlEnabled;
+    switch (rovstate) {
+    case RS_SLEEP:
+        send_buf[19] |= 0x02;
+        break;
+    case RS_SERVICE:
+        send_buf[19] |= 0x04;
+        break;
+    case RS_KEEPALIVE:
+        send_buf[19] |= 0x08;
+        break;
+    case RS_MISSION:
+        send_buf[19] |= 0x10;
+        break;
+    case RS_FAULT:
+        send_buf[19] |= 0x20;
```

Comment:
In the enum definition you can define each of these values to the value you're or-ing, e.g.:
```
RS_SLEEP = 0x02,
RS_SERVICE = 0x04,
[...etc...]
```
and then you could just do `send_buf[19] |= rovstate` here instead of the whole switch block

Reply:
Actually I guess you could just do `send_buf[19] = rovstate` in place of `send_buf[19] = 0`